Include more info the rustc invocation/compilation

Xanewok opened the pull request:
This is a proposal for what we might include about a given crate compilation. The most important one is the invocation, which can be used to possibly recreate a build plan of sorts from a set of save analysis JSON files. However, there are also other options included that might be handy, such as info whether this is a unit test harness mode, used sysroot, target triple and also where are the actual external crate files.

Additionally, I think it might also be useful to include `cfg` options (which I did at first, but decided that it's not as crucial now and can possibly be added later), which might be useful for the RLS and other consumers.

I'll also push a PR against Rust repo shortly, referencing this branch.
@nrc is this okay for now?

Xanewok committed:
Add CompilationOptions

nrc commented on src/lib.rs:
```diff
@@ -101,6 +103,20 @@ pub struct GlobalCrateId {
+pub struct CrateSource {
+    pub dylib: Option<String>,
+    pub rlib: Option<String>,
+    pub rmeta: Option<String>,
```
What exactly goes in these paths? Having the full path doesn't seem useful since we'd need to relocate, so I guess just the filenames. Will these always be the same when built on a different machine?

Why do we need to know what kind of crate was used?

Can there ever be more than one of these per crate (i.e., could this be an enum)?

nrc commented on src/lib.rs:
```diff
@@ -115,6 +131,17 @@ pub struct SpanData {
+#[derive(Debug, Clone)]
+pub struct CompilationOptions {
+    pub invocation: String,
+    pub crate_name: Option<String>,
```
Why is this an option?

nrc commented on src/lib.rs:
```diff
@@ -115,6 +131,17 @@ pub struct SpanData {
+    pub invocation: String,
+    pub crate_name: Option<String>,
+    pub test: bool,
+    pub sysroot: Option<PathBuf>,
```
how is this useful?

nrc commented on src/lib.rs:
```diff
@@ -115,6 +131,17 @@ pub struct SpanData {
+pub struct CompilationOptions {
+    pub invocation: String,
+    pub crate_name: Option<String>,
+    pub test: bool,
```
This seems a bit ad hoc. Would it be better to have a 'target' field or something more generic?

nrc commented on src/lib.rs:
```diff
@@ -101,6 +103,20 @@ pub struct GlobalCrateId {
+/// must be non-None.
+///
+/// Mimicks CrateSource from src/librustc/middle/cstore.rs (only paths)
+/// TODO: Uses String to be easily debuggable, change to PathBuf
```
Presumably this should be addressed.

nrc commented on src/lib.rs:
```diff
@@ -101,6 +103,19 @@ pub struct GlobalCrateId {
+    pub dylib: Option<PathBuf>,
+    pub rlib: Option<PathBuf>,
+    pub rmeta: Option<PathBuf>,
+}
```
I know we discussed this, but I'm not sure if we got to any kind of conclusion. IIRC, there's only ever multiple sources for std and we only want this info to avoid parsing the command line, but even then it would be better to use the crate name rather than the path. Is that right? Is there anything we need to make the disctinction between these kinds of crate sources?

Xanewok replied:
The main use for this was to figure out the dependencies between `rustc` compilations, combined with the new `output` field, so we can reconstruct file-based input-output 'white boxes' in the build plan.

However, we might just discard this information completely since crate disambiguator (included for the local crate and deps) information is sufficient to figure out the dependency relation between crates and by adding the `output` field we also get the implicit file-based dependencies.

I'll remove the struct in the next commit.

Xanewok commented:
Changed to include only 
```rust
pub struct CompilationOptions {
    pub directory: PathBuf,
    pub program: String,
    pub arguments: Vec<String>,
    pub output: PathBuf,
}
```
Also split `command` into `program` and `arguments` to make more sense semantically.
